File: app/repositories/agents.py
"""Repositório de agentes — encapsula todo o SQL da tabela `agents`."""

from app.db import read_connection, transaction
from app.domain import AgentConfig

# Campos editáveis via UPDATE (whitelist — evita update de colunas indevidas).
_EDITABLE = (
    "name", "system_prompt", "business_rules", "max_turns",
    "product_mode", "product_api_url", "product_api_key",
)


class AgentRepository:
    def get(self, agent_id: str) -> AgentConfig | None:
        with read_connection() as conn:
            row = conn.execute("SELECT * FROM agents WHERE id = ?", (agent_id,)).fetchone()
            return AgentConfig.from_row(row) if row else None

    def list(self) -> list[AgentConfig]:
        with read_connection() as conn:
            rows = conn.execute("SELECT * FROM agents ORDER BY created_at").fetchall()
            return [AgentConfig.from_row(r) for r in rows]

    def exists(self, agent_id: str) -> bool:
        with read_connection() as conn:
            return conn.execute("SELECT 1 FROM agents WHERE id = ?", (agent_id,)).fetchone() is not None

    def insert(self, agent_id: str, api_key: str, data: dict) -> None:
        with transaction() as conn:
            conn.execute(
                "INSERT INTO agents (id, name, api_key, system_prompt, business_rules, "
                "max_turns, product_mode, product_api_url, product_api_key) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                (
                    agent_id, data["name"], api_key,
                    data.get("system_prompt", ""), data.get("business_rules", ""),
                    data.get("max_turns", 15), data.get("product_mode", "none"),
                    data.get("product_api_url", ""), data.get("product_api_key", ""),
                ),
            )

    def update(self, agent_id: str, changes: dict) -> None:
        fields = {k: v for k, v in changes.items() if k in _EDITABLE and v is not None}
        if not fields:
            return
        sets = ", ".join(f"{k} = ?" for k in fields)
        with transaction() as conn:
            conn.execute(
                f"UPDATE agents SET {sets} WHERE id = ?",
                (*fields.values(), agent_id),
            )

    def delete(self, agent_id: str) -> bool:
        with transaction() as conn:
            return conn.execute("DELETE FROM agents WHERE id = ?", (agent_id,)).rowcount > 0

File: src/server/repositories/agents.ts
/**
 * Repositório de agentes (porta `app/repositories/agents.py`).
 *
 * INVARIANTE DE TENANCY: resolução por (tenant_id, slug); listagens sempre
 * escopadas por tenant. A PK `id` é opaca e global ({tenant}__{slug}).
 */

import { and, asc, eq } from "drizzle-orm";

import { db } from "../db/client";
import { agents } from "../db/schema";
import type { AgentConfig, ProductMode } from "../domain";
import type { AgentCreateData, AgentUpdateData } from "../schemas";
import { toIso } from "./util";

type AgentRow = typeof agents.$inferSelect;

function toAgentConfig(r: AgentRow): AgentConfig {
  return {
    id: r.id,
    tenantId: r.tenantId,
    slug: r.slug,
    name: r.name,
    systemPrompt: r.systemPrompt ?? "",
    businessRules: r.businessRules ?? "",
    maxTurns: r.maxTurns,
    productMode: r.productMode as ProductMode,
    productApiUrl: r.productApiUrl ?? "",
    productApiKey: r.productApiKey ?? "",
    ragEnabled: r.ragEnabled,
    externalProducts: r.externalProducts,
    skills: Array.isArray(r.skills) ? r.skills : [],
    createdAt: toIso(r.createdAt),
  };
}

export class AgentRepository {
  async getById(agentId: string): Promise<AgentConfig | null> {
    const rows = await db.select().from(agents).where(eq(agents.id, agentId)).limit(1);
    return rows[0] ? toAgentConfig(rows[0]) : null;
  }

  async get(tenantId: string, slug: string): Promise<AgentConfig | null> {
    const rows = await db
      .select()
      .from(agents)
      .where(and(eq(agents.tenantId, tenantId), eq(agents.slug, slug)))
      .limit(1);
    return rows[0] ? toAgentConfig(rows[0]) : null;
  }

  async listForTenant(tenantId: string): Promise<AgentConfig[]> {
    const rows = await db
      .select()
      .from(agents)
      .where(eq(agents.tenantId, tenantId))
      .orderBy(asc(agents.createdAt));
    return rows.map(toAgentConfig);
  }

  async listAll(): Promise<AgentConfig[]> {
    const rows = await db.select().from(agents).orderBy(asc(agents.createdAt));
    return rows.map(toAgentConfig);
  }

  async exists(tenantId: string, slug: string): Promise<boolean> {
    const rows = await db
      .select({ id: agents.id })
      .from(agents)
      .where(and(eq(agents.tenantId, tenantId), eq(agents.slug, slug)))
      .limit(1);
    return rows.length > 0;
  }

  async insert(agentId: string, tenantId: string, slug: string, data: AgentCreateData): Promise<void> {
    await db.insert(agents).values({
      id: agentId,
      tenantId,
      slug,
      name: data.name,
      systemPrompt: data.systemPrompt,
      businessRules: data.businessRules,
      maxTurns: data.maxTurns,
      productMode: data.productMode,
      productApiUrl: data.productApiUrl,
      productApiKey: data.productApiKey,
      ragEnabled: data.ragEnabled,
      externalProducts: data.externalProducts,
      skills: data.skills,
    });
  }

  async update(agentId: string, changes: AgentUpdateData): Promise<void> {
    if (Object.keys(changes).length === 0) return;
    await db
      .update(agents)
      .set(changes as Partial<typeof agents.$inferInsert>)
      .where(eq(agents.id, agentId));
  }

  async delete(agentId: string): Promise<boolean> {
    const deleted = await db.delete(agents).where(eq(agents.id, agentId)).returning({ id: agents.id });
    return deleted.length > 0;
  }
}
